### PR TITLE
[relayd] add session timeout

### PR DESCRIPTION
**Pull request by fbrendel**
Configure a timeout value for applications that idles longer than the default of 10 minutes.

### CODE REVIEW

File: net/relayd/src/www/load_balancer_virtual_server_edit.php [hunk 1]
```diff
@@ -100,6 +100,10 @@
         $input_errors[] = gettext("You cannot select a Fall Back Pool when using the DNS relay protocol.");
     }
 
+    if ($pconfig['sessiontimeout'] != "" && ($pconfig['sessiontimeout'] < 1 || $pconfig['sessiontimeout'] > 2147483647)) {
```
Comment (fichtner) [inline net/relayd/src/www/load_balancer_virtual_server_edit.php]: must improve this check: !empty() and if integer...

**Comment by fbrendel** [comment]
Hi Franco,
and if you ask the '> 2147483647' test is for 64bit environments. :smiley:

Kind regards
Frank

File: net/relayd/src/www/load_balancer_virtual_server_edit.php [hunk 2]
```diff
@@ -100,6 +100,15 @@
         $input_errors[] = gettext("You cannot select a Fall Back Pool when using the DNS relay protocol.");
     }
 
+    if (isset($pconfig['sessiontimeout']) &&
+            (!is_numeric($pconfig['sessiontimeout']) ||
+              $pconfig['sessiontimeout'] < 1 ||
+              $pconfig['sessiontimeout'] > 2147483647)) {
+        $input_errors[] = gettext("The session timeout value must be an integer between 1 and 2147483647.");
+    } else {
+    	$pconfig['sessiontimeout'] = intval($pconfig['sessiontimeout']);
```
Comment (fichtner) [inline net/relayd/src/www/load_balancer_virtual_server_edit.php]: okay, letzte Frage: brauchen wir das überhaupt, ist doch hier nur validierung? :)

Comment (fbrendel) [inline net/relayd/src/www/load_balancer_virtual_server_edit.php]: Hmm, wer prüft das sonst und warum sind die anderen Checks drin wenn sie nicht gebraucht werden?

Comment (fbrendel) [inline net/relayd/src/www/load_balancer_virtual_server_edit.php]: Also dieser Request zieht sich ganz schön hin... :walking_man: :smile:

Comment (fabianfrz) [inline net/relayd/src/www/load_balancer_virtual_server_edit.php]: @fbrendel nur kein Stress. Das nächste release gibt es eh erst nächste Woche.

File: net/relayd/src/www/load_balancer_virtual_server_edit.php [hunk 3]
```diff
@@ -314,6 +323,16 @@
                       </td>
                     </tr>
                     <tr>
+                    <tr>
+                      <td><a id="help_for_sessiontimeout" href="#" class="showhelp"><i class="fa fa-info-circle"></i></a> <?=gettext("Session Timeout"); ?></td>
+                      <td>
+                        <input type="text" name="sessiontimeout" id="sessiontimeout" value="<?=$pconfig['sessiontimeout'];?>" />
+                        <div class="hidden" for="help_for_sessiontimeout">
+                          <?=gettext("This is the timeout in seconds for idle sessions."); ?>
+                          <br /><?=gettext(" The default timeout is 600, the maximum is 2147483647 (68 years)."); ?>
```
Comment (fabianfrz) [inline net/relayd/src/www/load_balancer_virtual_server_edit.php]: Could you please remove this space at the beginning of the line?

Comment (fabianfrz) [inline net/relayd/src/www/load_balancer_virtual_server_edit.php]: replace with:
```
<?= sprintf(gettext("The default timeout is %s, the maximum is %s (%s years)."), '600', '2147483647', '68') ?>
```

File: net/relayd/src/www/load_balancer_virtual_server_edit.php (hunk 2 above)
Comment (fbrendel) [inline net/relayd/src/www/load_balancer_virtual_server_edit.php]: @fichtner wäre vermutl. schneller gewesen und hätte weniger Arbeit gehabt.
Immerhin krieg ich 'nen Lerneffekt :man_teacher:

Comment (fbrendel) [inline net/relayd/src/www/load_balancer_virtual_server_edit.php]: Schade, hat jetzt doch nicht mit 17.1.5 geklappt.
Fehlt da noch was?

**Comment by fichtner** [comment]
will get this in for 17.1.6 :)

**Comment by fichtner** [comment]
@fbrendel If you get a chance to test this please do. The validation of "", "0", "00001" and so on is very silly in PHP at times so I had to add more stuff.

**Comment by fbrendel** [comment]
@fichtner Ok, tested with the following values
0 - shows error
1 - ok
001 - converts to 1
"001" - shows error
'' - shows error
1234567890987654321 - shows error
-01 - shows error
-1 - shows error
+1 - converts to 1
တစ် - shows error

If this is not enough we could block any keys via javascript that are not numbers. :smirk:

**Comment by fbrendel** [comment]
And I would add a '_and less than 2147483647_' to the error message.

**Comment by fichtner** [comment]
Thanks for testing, it's good enough. 😊

I removed the number from the validation error, the risk of the value being set is next to none, likely a config error. 32 bit time also gets in the way before this would matter.

**Pull request closed without merging**